### ISSUE TITLE
[question] Enrollment codes for BeNext/Zipato Keypad

### ISSUE THREAD

**Issue by lolautruche** [zwave-js/zwavejs2mqtt#782]
Hello

Sorry in advance if this question has already been posted/answered.
I have a BeNext/Zipat mini-keypad that I used to use with Domoticz/OZW. I am now using ZWaveJS2mqtt with HomeAssistant, and I'd like to use my tags. I can see in the UI that I can manually add user codes (most likely including the tags codes), but I can't figure out how to enroll my tags. I tried to press home/away on my tag reader using debug, but the code won't appear here.
I currently don't have a compatible NFC reader, but would buy one if needed.

Can you please help me?

Thanks in advance

**Comment by robertsLando** [comment]
@lolautruche Could you attach a gif showing what you are actually try to do on the UI and also both z2m and zwavejs logs?

**Comment by lolautruche** [comment]
Hi @robertsLando 

On the UI, I couldn't figure out how to do or what value to inject in the "user codes" section.

I checked if the user code from my badge would appear in the logs but I had no luck.

Here are the logs from z2m:

```
2021-03-02 13:58:55.232 INFO ZWAVE: Node 3: value updated: 112-0-4 1 => 1
2021-03-02 13:59:02.853 DEBUG MQTT: Publishing to zwave/Keypad/99/0/userIdStatus: { time: 1614693542852, value: 0 } with options { qos: 1, retain: true }
2021-03-02 13:59:02.854 INFO ZWAVE: Node 25: value updated: 99-0-userIdStatus-0 0 => 0
2021-03-02 13:59:02.857 DEBUG MQTT: Publishing to zwave/Keypad/99/0/userCode: { time: 1614693542856, value: <Buffer 8f 1f 35 ad 29 00 01 04> } with options { qos: 1, retain: true }
2021-03-0
```

With another badge:

```
2021-03-02 14:04:24.089 DEBUG MQTT: Publishing to zwave/Keypad/99/0/userIdStatus: { time: 1614693864088, value: 0 } with options { qos: 1, retain: true }
2021-03-02 14:04:24.090 INFO ZWAVE: Node 25: value updated: 99-0-userIdStatus-0 0 => 0
2021-03-02 14:04:24.092 DEBUG MQTT: Publishing to zwave/Keypad/99/0/userCode: { time: 1614693864091, value: <Buffer 8f 05 e3 87 27 00 01 04> } with options { qos: 1, retain: true }
2021-03-02 14:04:24.093 INFO ZWAVE: Node 25: value updated: 99-0-userCode-0 ��' => ��'
```

Please note that I included the tag reader with secure inclusion.

The zwavejs log doesn't have relevant information:

```
13:59:02.870 CNTRLR « [Node 025] received wakeup notification
14:04:24.109 CNTRLR « [Node 025] received wakeup notification
```

**Comment by robertsLando** [comment]
@lolautruche Please attach to your issue all logs needed: https://zwave-js.github.io/zwavejs2mqtt/#/troubleshooting/bug_report?id=bug-report

> but I can't figure out how to enroll my tags. I tried to press home/away on my tag reader using debug, but the code won't appear here

I never used those devices but I think a manual would be useful in this case. BTW @AlCalzone said to me that other users having the same devices had no issues using them

**Comment by lolautruche** [comment]
[zwavejs_1.log](https://github.com/zwave-js/zwavejs2mqtt/files/6069327/zwavejs_1.log)
[zwavejs2mqtt.log](https://github.com/zwave-js/zwavejs2mqtt/files/6069328/zwavejs2mqtt.log)

**Comment by lolautruche** [comment]
If others can use it, then I'm pretty sure I can :-)
It's just that I don't understand how I can add the code of my badges as I don't know them, unfortunately. And I'm not sure what value to add (I'm suspecting this should be hex value)

**Comment by robertsLando** [comment]
@lolautruche What version are you using? Could you update to latest please? 2.1.0 ATM

**Comment by AlCalzone** [comment]
> zwavejs_1.log

That's not loglevel debug nor silly 🤷🏻‍♂️ 

Afaik the new codes should appear as user code 0 or something and you have to manually copy them to another user code where you want to use them.

**Comment by lolautruche** [comment]
Oh my... Sorry, I was convinced that I set it to debug. Let me try again :-)

**Comment by lolautruche** [comment]
@robertsLando 
<img width="825" alt="Capture d’écran 2021-03-02 à 15 36 47" src="https://user-images.githubusercontent.com/313528/109664184-2d80ce80-7b6d-11eb-9296-b8ce5daa0b3a.png">

I did a `docker-compose pull` this morning.

**Comment by robertsLando** [comment]
app version 1.0.5, latest is 2.1.0, I think something isn't wrong in your docker compose, what tag are you using? latest?

**Comment by lolautruche** [comment]
Weird...
I'm indeed using the `latest` tag.
Here is my `docker-compose.yml`

```yaml
version: "3"
services:
  homeassistant:
    container_name: home-assistant
    image: homeassistant/home-assistant:stable
    volumes:
      - /home/lolautruche/homeassistant/config:/config
    devices: []
      #- '/dev/ttyUSB0:/dev/ttyUSB0'
      #- /dev/ttyUSB1:/dev/ttyUSB1
      #- /dev/ttyACM0:/dev/ttyACM0
    environment:
      - TZ=Europe/Paris
    restart: unless-stopped
    network_mode: host

  zwavejs2mqtt:
    image: zwavejs/zwavejs2mqtt:latest
    restart: unless-stopped
    tty: true
    stop_signal: SIGINT
    #networks:
    #  - zwave
    devices:
      - '/dev/ttyACM0:/dev/ttyACM0'
    volumes:
      - ./zwavejs/store:/usr/src/app/store
    ports:
      - '8091:8091' # port for web interface
      - '3000:3000' # port for zwave-js websocket server

  mosquitto:
    image: eclipse-mosquitto
    ports:
      - '1883:1883'
      - '9001:9001'
    restart: unless-stopped
    volumes:
      - ./mqtt/config/mosquitto.conf:/mosquitto/config/mosquitto.conf
      - ./mqtt/data:/mosquitto/data
      - ./mqtt/log:/mosquitto/log

  portainer:
    image: portainer/portainer-ce
    volumes:
      - portainer_data:/data
      - /var/run/docker.sock:/var/run/docker.sock
    ports:
      - 8000:8000
      - 9000:9000
    restart: unless-stopped

volumes:
  portainer_data:
```

**Comment by lolautruche** [comment]
@AlCalzone I indeed figured out using silly log level that the user code does appear in the log (see attached). In the UI I have the following:

<img width="759" alt="Capture d’écran 2021-03-02 à 15 44 06" src="https://user-images.githubusercontent.com/313528/109665175-31f9b700-7b6e-11eb-911c-1f1d8c6ec12a.png">
[zwavejs_1(1).log](https://github.com/zwave-js/zwavejs2mqtt/files/6069405/zwavejs_1.1.log)

**Comment by lolautruche** [comment]
[zwavejs_1.log](https://github.com/zwave-js/zwavejs2mqtt/files/6069447/zwavejs_1.1.log)

**Comment by AlCalzone** [comment]
With the newest zwavejs2mqtt (I think @robertsLando is doing one more change though), the `[object ArrayBuffer]` should become a HEX string which you can copy to the desired user code #

**Comment by robertsLando** [comment]
@lolautruche could you click on the button `Dump` in the top of control panel nodes table and attach here the nodes dump?

**Comment by AlCalzone** [comment]
Oh and you should probably re-interview Node 25, zwave-js couldn't read the user codes so far.
I've deleted your previous log since it contains the user code you tried to teach in.

**Comment by lolautruche** [comment]
@AlCalzone That's cool, thanks!
@robertsLando do you mean the "export" button?

**Comment by robertsLando** [comment]
![Schermata da 2021-03-02 15-53-28](https://user-images.githubusercontent.com/11502495/109666447-79cd0e00-7b6f-11eb-9708-e0afdd38df00.png)

The yellow one, `DUMP`

**Comment by lolautruche** [comment]
> Oh and you should probably re-interview Node 25, zwave-js couldn't read the user codes so far.
> I've deleted your previous log since it contains the user code you tried to teach in.

Yes I know, it seems that there was an unexpected restart or something. This occurs quite a lot these days.

@robertsLando How can it be that I'm using the `latest` tag but I'm still getting `1.0.5`? I can switch to the `2.1.0` tag, but still I don't understand.

**Comment by lolautruche** [comment]
I don't have this yellow button. Probably because of the version

**Comment by lolautruche** [comment]
OK, I was able to force the update. Let me try again...

**Comment by robertsLando** [comment]
Remember to re-interview your device (by sending re-interview action)

**Comment by lolautruche** [comment]
[nodes_dump.json.gz](https://github.com/zwave-js/zwavejs2mqtt/files/6069548/nodes_dump.json.gz)

**Comment by robertsLando** [comment]
@lolautruche Please try to use `master` tag, just wait 30 mins until https://github.com/zwave-js/zwavejs2mqtt/runs/2014457862?check_suite_focus=true ends

**Comment by lolautruche** [comment]
Awesome thank you

**Comment by lolautruche** [comment]
Now that I have the user code in hex (thanks to the logs), I tried to set the first user code "slot" and I have this message:

```
Driver: UserCodeCCSet: The user code must have a length of 4 to 10 characters
```

**Comment by AlCalzone** [comment]
well, what did you write in there?

**Comment by lolautruche** [comment]
The hex code.
I know I didn't wait, I tried before getting the message from @robertsLando ^^

**Comment by robertsLando** [comment]
@lolautruche Available now

**Comment by lolautruche** [comment]
Thank you

However, I don't see anything now 😭 : 
<img width="718" alt="image" src="https://user-images.githubusercontent.com/313528/109677040-8d7d7200-7b79-11eb-8a1f-9477416db5ae.png">

**Comment by robertsLando** [comment]
Ops, I fixed it on backend side but forgot to add the parsing on ui side, it's a matter of a line of code, let me fix this (but you will have to wait new build)

**Comment by lolautruche** [comment]
No problem, I understand 😄 
What is supposed to appear is a string like `0x0000000000000000`, right? Because, even by doing so, I'm still getting the error I mentioned above:

```
Driver: UserCodeCCSet: The user code must have a length of 4 to 10 characters
```

**Comment by lolautruche** [comment]
This occurs when I select "Enabled" and then I click send. Beforehand I paste my hex code in the "user code" field + send

**Comment by AlCalzone** [comment]
That is an 8-byte buffer as hex, so it should work if zwavejs2mqtt converts it correctly

**Comment by robertsLando** [comment]
Wait until this ends: https://github.com/zwave-js/zwavejs2mqtt/actions/runs/614731828 then repull `master`

**Comment by lolautruche** [comment]
It's back on "[object ArrayBuffer]"

<img width="392" alt="image" src="https://user-images.githubusercontent.com/313528/109688713-dedf2e80-7b84-11eb-81ec-9fbc1876bc43.png">

**Comment by lolautruche** [comment]
nvm, I finally got the user code appearing correctly. Not sure why I had the ArrayBuffer string though.
Now I have to figure out how to use the keypad with Home Assistant, as I don't see any value updating changing using MQTT or the ZwaveJS integration.

Thanks a lot for your help=

**Comment by robertsLando** [comment]
You can easily integrate them in this way: 

https://zwave-js.github.io/zwavejs2mqtt/#/guide/homeassistant?id=manually-create-entities

I also suggest you to use mqtt explorer to find the usercode topics and make some tries with the payload to send

---

Daniel

**Comment by lolautruche** [comment]
Thank you so much @robertsLando
